### ISSUE TITLE
JOJI support [Request]

### ISSUE THREAD

**Issue by Roadhog360** [libertyernie/BrawlManagers#14]
Support for JOJI's music list modifiers would be great, as well as reading and modifying names set with his codes and outputting these codes to both the .txt or .gct (or both at the same time) would be fantastic, would make changing my entries a LOT easier.

**Comment by libertyernie** [comment]
Could you post a link to the code?

**Comment by Roadhog360** [comment]
http://ssbbhack.web.fc2.com

There are a few codes. Their use is badly explained, so if you don't understand anything, I'll explain to you personally. (If you got a Discord account and added me at Roadhog360#7777, it'd be easier to show you)

Before starting, make sure you have the soundbank expansion code. It's not by JOJI, but is very common among modders. I believe you need v3 of the expansion code.
Go down the side bar and select "SSBB CODES" from tab.
Then, find the codes "CMM:Custom My Music V2 [JOJI]", "Tracklist Modifier v1.0 (For CMM:Custom My Music)".

Now, we'll also need "CMM Setting Code [JOJI]" to configure My Music. Be sure to copy one for every stage into your codeset, and also copy the ones for the custom stages you use.

If you want to add completely new music slots, you can, even with names!  Copy "CNMT:Custom New Music Title (RSBE.Ver)" into your codeset. This allows the game to load your custom song titles. Now, "CNMT Setting Code [JOJI]" is what you need to configure.
Back on the code list in the top of the page, click "CNMT:Custom New Music Title for CSE v3.0 [JOJI]" to get the track IDs.

That should be all. If it's badly explained on the website and you don't understand, hop on Discord and I'll help.

**Comment by libertyernie** [comment]
The CMM Setting Code looks fairly straightforward to me. I don't think I'll add support for editing the code, but I would like to support reading it if it's there.

Is the CNMT Setting Code for letting you define song names in your .gct codeset instead of editing info.pac?

**Comment by Roadhog360** [comment]
Technically yes, but it's also for adding new song names, which info.pac does not support.

**Comment by Roadhog360** [comment]
What do you mean "Assumes you do not mess with JOJI's tracklist modifier"?
What happens if you have, because I have.

**Comment by libertyernie** [comment]
I haven't written any code yet that reads the tracklist modifier. I still
have to do that.

Could you post the tracklist modifier code you're using?

**Comment by Roadhog360** [comment]
I've modified mine, so you won't see the default lists. JOJI has tracklist modifiers with all of the default song lists in them.
If you still want my modifiers, it'll be easier to just give you the entire codeset. If you want that, please give me a way to privately contact you to give you the .gct and .txt.

**Comment by libertyernie** [comment]
Does yours start with the same bytes as joji's? If so I can just use the one they posted to test.

    0610F9BC 00000014
    60000000 60000000
    3C80811A 7C601B78
    608447E7 00000000
    4A000000 81000000
    1417DF44 7CE03B78
    1417DF4C 7C0400AE
    301A47E8 00000000
    161A47E8 000000LL (LL = bytes of following data)

**Comment by Roadhog360** [comment]
No, I upped that value as I added a lot of new song slots.
There are exactly 3 stage song lists I haven't touched, and those are the only three values that remain as default.
Again, JOJI has all of the game's default tracklists preconfigured as codes.

**Comment by Roadhog360** [comment]
http://ssbbhack.web.fc2.com/codersbe.html#CMML
Here, scroll down for default tracklist codes for all stages.

**Comment by libertyernie** [comment]
I just uploaded a new version. Let me know how it works.

**Comment by Roadhog360** [comment]
Sorry for the delay.
The Stage Manager now just gives an unhandled exception whenever I click a stage and has no music list. 
The Song Manager will just not work at all. I open it, and nothing happens, it doesn't open.

**Comment by Roadhog360** [comment]
I can send you my codeset txt and gct if you need to look at my code set.

**Comment by libertyernie** [comment]
That would be helpful. My email address is

**Comment by Roadhog360** [comment]
Alright, I'll send it in a bit. I'd recommend editing out that now to prevent any people from getting funny ideas like spamming you.

**Comment by libertyernie** [comment]
Quick question: does the Sector Z music (28A6) show up as "Sector Z64" or "Sector Z 64"? It looks like it appears twice in the code, and I'm wondering whether the first or second appearance takes precedence.

I think this is also the cause of the crash, by the way. But if it works fine in Brawl there's no reason my program shouldn't handle it.

Also, you have song E51F on Halberd, right? Just want to check that it's reading the code right.

**Comment by Roadhog360** [comment]
Huh, that must have been a mistake.
Sector Z  64 is what shows up, I don't remember entering it twice.
What line is this on?
I also had Halberd's music listed twice as well. Yes. that song is on Halberd.

**Comment by Roadhog360** [comment]
Okay, I found and fixed the errors. Song Manager still won't boot and I still get an exception whenever I choose a stage in Stage Manager.
Stage Manager doesn't crash, it just gives an exception when I pick a stage and will not display any music in its list.
I should also mention that Song Manager DOES appear briefly as a background process in Task Manager before closing.

**Comment by libertyernie** [comment]
I think I'll be able to fix those issues in the next version.

**Comment by Roadhog360** [comment]
Oh, nice. I take it even after fixing the errors we both found in the codeset, you were able to reproduce both bugs I brought up?

**Comment by libertyernie** [comment]
Yeah, I got them both to open without crashing and display the song titles.

**Comment by Roadhog360** [comment]
Hm... but what about before any attempted fix? Did you get the issues I was describing?

Oh, and are these applications able to change the code in the .txt and/or .gct?

**Comment by libertyernie** [comment]
Yeah, i saw the same issues.

These programs can't edit any of the JOJI codes, only read them.

**Comment by libertyernie** [comment]
I just posted version 4.1 - try it out and let me know how it works.

**Comment by Roadhog360** [comment]
Alright. Out of curiosity, do you ever plan on adding support for writing to the codes, and saving it to the .gct and .txt?

**Comment by Roadhog360** [comment]
Alright, well, it seems the issues are fixed. The Stage Manager shows the song lists and Song manager displays all of my music with the correct title.
However, there's one more error with the Stage Manager with the song list; that wasn't previously seen since it errored before. Now, any JOJI song on the Stage Manager will not display its title. Song Manager does not have this issue, so all you need to worry about is the Stage Manager.

On a much more minor note, there appears to be a good lag spike between switching stages in the manager. The application seems to freeze completely for about a second when choosing a new stage.

**Comment by libertyernie** [comment]
Does it just show an empty song title?

When I go to STGCUSTOM09, it shows 0000E56F as the first song, and shows its title as Sky Sanctuary Zone (although I can't edit it.)

I don't plan on adding edit support for these codes at the moment.

**Comment by Roadhog360** [comment]
For me, Stage Manager does not show the custom song title, it is empty, yes. It only shows them on Song Manager.

**Comment by libertyernie** [comment]
What if you manually load your GCT codeset from the File menu? If that fixes it, it means I need to fix Stage Manager's detecting of the GCT path (relative to the stage/melee folder). If it doesn't, I'll have to investigate further.

**Comment by Roadhog360** [comment]
It seems that works.

**Comment by Roadhog360** [comment]
I should note I have BrawlManagers directly on my SD card for convenience since it seems to always auto-detect my stuff there.

**Comment by Roadhog360** [comment]
Oh, StageBox fixed an issue where there's static when playing BRSTMs, you might want to apply this to your audio players, I can still hear static when playing my BRSTM files when using BrawlManagers.

**Comment by libertyernie** [comment]
What folder is your RSBE01.gct in on your SD card? Is it in /codes? If not I probably need to add something.

**Comment by Roadhog360** [comment]
Yes.

**Comment by libertyernie** [comment]
Hmm, mine is in /codes too but it detects that fine.

I'm trying to reproduce your problem because if I can't, then I don't know whether I've fixed it or not :/

I can at least apply the StageBox update, though. I forgot last time that it didn't include BrawlBox's BRSTM fixes, but it does now.

**Comment by Roadhog360** [comment]
Stage Manager minor bug: Can't see the .brstm name of JOJI songs; the list is way too thin and I can't make it thicker, and it covers the file name.
Maybe making it so the first four zeros don't appear?
(Deleted 'image upload' text since the upload somehow failed.)

**Comment by Roadhog360** [comment]
![image](https://user-images.githubusercontent.com/11903511/43684285-f135227c-9862-11e8-94b8-fbbce0509af2.png)
For some reason that messed up.

**Comment by libertyernie** [comment]
This is a pretty simple fix. I'll fix it next time there's an update.

**Comment by Roadhog360** [comment]
![image](https://user-images.githubusercontent.com/11903511/43986993-c623366a-9cdf-11e8-97eb-799723f9636c.png)
Can this mode in Song Manager support custom stages and custom stage lists? I need an easy way to get all my song names and brstm names all at once and this would make it very easy.

**Comment by Roadhog360** [comment]
I just realised I've been reusing this issue instead of opening new ones. Oops. I'm going to close this and post anything else in a new issue.
